### PR TITLE
[feature](nereids) Support analyze for test SSB

### PR DESCRIPTION
**Pull request by 924060929**
# Proposed changes

Issue Number: no issue

## Problem Summary:

Follow-up #10241, this PR go through parse and analyze the SSB and add this functions:
1. support parse parenthesizedExpression
2. support analyze LogicalAggregate and LogicalSort
3. replace the functionCall to UnboundFunction and BoundFunction
4. support sum aggregate funciton
5. fix the dead loop in the ExpressionRewriter
6. refine some code

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
7. Does it need to update dependencies: (Yes/No)
8. Are there any changes that cannot be rolled back: (Yes/No)

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -143,265 +130,47 @@ public Object visitChildren(RuleNode node) {
...
-            withAgg = withAggClause(namedExpressions, aggClause.groupByItem(), withFilter);
-        } else {
-            withAgg = withProject;
-        }
-
-        return withAgg;
-    }
-
-    @Override
-    public LogicalPlan visitFromClause(FromClauseContext ctx) {
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:269]: okey

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -143,265 +130,47 @@ public Object visitChildren(RuleNode node) {
...
+        });
     }
 
     @Override
     public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationContext ctx) {
-        Supplier<LogicalPlan> f = () -> {
+        return ParserUtils.withOrigin(ctx, () -> {
             // TODO: support on row relation
-            LogicalPlan from = visitFromClause(ctx.fromClause());
+            LogicalPlan relation = withRelation(Optional.ofNullable(ctx.fromClause()));
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:153]: okey

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundAlias.java [hunk 3]
```diff
@@ -34,18 +38,28 @@
...
+    public UnboundAlias(ExprId exprId, CHILD_TYPE child) {
         super(NodeType.UNBOUND_ALIAS, child);
+        this.exprId = Objects.requireNonNull(exprId, "exprId can not be null");
+        this.name = "expr_" + exprId;
     }
 
     @Override
     public String sql() {
-        return null;
+        return child().sql();
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundAlias.java:52]: okey

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundAlias.java [hunk 4]
```diff
@@ -34,18 +38,28 @@
         extends NamedExpression
         implements UnaryExpression<CHILD_TYPE>, Unbound {
 
-    public UnboundAlias(CHILD_TYPE child) {
+    private final ExprId exprId;
+    private final String name;
+
+    public UnboundAlias(ExprId exprId, CHILD_TYPE child) {
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundAlias.java:44]: okey

**Review by EmmyMiao87 — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ProjectToGlobalAggregate.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.functions.AggregateFunction;
+import org.apache.doris.nereids.trees.plans.Plan;
+
+import com.google.common.collect.ImmutableList;
+
+/** ProjectToGlobalAggregate. */
+public class ProjectToGlobalAggregate extends OneAnalysisRuleFactory {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ProjectToGlobalAggregate.java:30]: coool~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/TreeNode.java [hunk 6]
```diff
@@ -46,4 +48,27 @@
 
     NODE_TYPE withChildren(List<NODE_TYPE> children);
 
+    default void foreach(Consumer<TreeNode<NODE_TYPE>> func) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/TreeNode.java:51]: this is a top-down foreach, so maybe it is better to add traverse order info into function name

**Review by morrySnow — COMMENTED** [review]
Great Job👏🏻

File: fe/fe-core/src/main/java/org/apache/doris/nereids/pattern/GroupExpressionMatching.java [hunk 7]
```diff
@@ -102,9 +104,12 @@ public GroupExpressionIterator(Pattern<Plan, Plan> pattern, GroupExpression grou
                     Group childGroup = groupExpression.child(i);
                     List<Plan> childrenPlan = matchingChildGroup(pattern, childGroup, i);
                     childrenPlans.add(childrenPlan);
+                    if (childrenPlans.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/pattern/GroupExpressionMatching.java:107]: yes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/TreeNode.java (hunk 6 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/TreeNode.java:51]: I think we have a promise: if the traverse function no order info, it must be top-down implicitly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 8]
```diff
@@ -50,50 +55,99 @@ public List<Rule<Plan>> buildRules() {
...
+            .flatMap(slot -> {
+                if (slot instanceof BoundStar) {
+                    return ((BoundStar) slot).getSlots().stream();
+                } else {
+                    return Stream.of(slot);
+                }
+            }).collect(Collectors.toList());
+    }
+
+    private <E extends Expression> List<E> bind(List<E> exprList, List<Plan> inputs, boolean isProjection) {
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:116]: I will replace the isProjection to Plan plan, and check `plan instanceof LogicalProject` in the visitUnboundStar()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindFunction.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    private <E extends Expression> List<E> bind(List<E> exprList) {
+        return exprList.stream()
+            .map(expr -> bind(expr))
+            .collect(Collectors.toList());
+    }
+
+    private <E extends Expression> E bind(E expr) {
+        return new FunctionBinder().bind(expr);
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindFunction.java:68]: okey

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 10]
```diff
@@ -114,35 +168,60 @@ public Slot visitUnboundSlot(UnboundSlot unboundSlot, Void context) {
...
+                    // table.*
+                    case 1:
+                        List<String> boundSlotQualifier = boundSlot.getQualifier();
+                        switch (boundSlotQualifier.size()) {
+                            // bound slot is `column` and no qualified
+                            case 0: return false;
+                            case 1: // bound slot is `table`.`column`
+                                return qualifierStar.get(0).equalsIgnoreCase(boundSlotQualifier.get(0));
+                            case 2:// bound slot is `db`.`table`.`column`
+                                return qualifierStar.get(0).equalsIgnoreCase(boundSlotQualifier.get(1));
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:199]: > what will happen?

sparksql will select all column in the db.t1 and the t1.
doris legacy analyzer will throw exception: Column 'xxx' in is ambiguous.

> what should happen?

compatible with previous behavior and throw exception.

---
I will add a fixme comment

File: fe/fe-core/src/main/java/org/apache/doris/nereids/pattern/GroupExpressionMatching.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/pattern/GroupExpressionMatching.java:107]: I add a ut

**Review by morrySnow — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ExpressionConverter.java [hunk 11]
```diff
@@ -134,16 +130,16 @@ public Expr visitLiteral(Literal literal, PlanTranslatorContext context) {
 
     // TODO: Supports for `distinct`
     @Override
-    public Expr visitFunctionCall(FunctionCall function, PlanTranslatorContext context) {
+    public Expr visitBoundFunction(BoundFunction function, PlanTranslatorContext context) {
         List<Expr> paramList = new ArrayList<>();
-        for (Expression expr : function.getFnParams().getExpressionList()) {
+        for (Expression expr : function.getArguments()) {
             paramList.add(visit(expr, context));
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ExpressionConverter.java:136]: DefaultExpressionVisitor.visit always returns null
```suggestion
            paramList.add(expr.accept(this, context));
```

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ExpressionConverter.java [hunk 12]
```diff
@@ -44,19 +45,14 @@
  * Used to convert expression of new optimizer to stale expr.
  */
 @SuppressWarnings("rawtypes")
-public class ExpressionConverter extends ExpressionVisitor<Expr, PlanTranslatorContext> {
+public class ExpressionConverter extends DefaultExpressionVisitor<Expr, PlanTranslatorContext> {
 
     public static ExpressionConverter converter = new ExpressionConverter();
 
     public static Expr convert(Expression expression, PlanTranslatorContext planContext) {
         return converter.visit(expression, planContext);
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ExpressionConverter.java:53]:
```suggestion
        return expression.accept(converter, planContext);
```

**Pull request merged**